<record-name>.constructor instead of recCon-NOT-PRINTED

andreasabel opened the pull request:
For #4697 I changed Agda to name anonymous record constructors as `R.constructor` where `R`is the name of the record type in question.

This change has some non-local effect on the the printer of bound variables.  
@jespercockx, could you take a look?

andreasabel committed:
[ #4697 ] rename recCon-NOT-PRINTED to <record>.constructor

Gives some more information for the crash described in #4697.

andreasabel committed:
[ #4697 ] test/Fail: mysterious change in printing bound variable

Renaming recCon-NOT-PRINTED into <record-name>.constructor mysteriously
changes the printing of bound variable A to A_1.

I am suspecting Jesper's stateful pretty printer to cause such spooky
actions at a distance...

jespercockx commented:
I'm not 100% sure what is causing the difference, but I think it might be because now the function `prettyTCM` is called, which will register the names in the context for shadowing, where this wasn't the case before.

Whatever is causing the change, the new error message is actually better than the old one, since it correctly disambiguates between the module parameter `A` and the implicit argument `A`. So I wouldn't worry about it too much.

andreasabel commented:
> now the function `prettyTCM` is called, which will register the names in the context for shadowing

Ok, this would be a way to fix the problem.

However, in general this means that debug printing (via `prettyTCM`) may interfere with ordinary printing?

> the new error message is actually better than the old one, since it correctly disambiguates between the module parameter `A` and the implicit argument `A`

Dunno, I liked the old error better, because it prints what the user wrote.  The user can still not see which of the `A`s has been renamed to `A_1`.

jespercockx commented:
> However, in general this means that debug printing (via prettyTCM) may interfere with ordinary printing?

Yes, that's an unfortunate (?) side effect of the new way I implemented the shadowing. Since it is only a problem when using verbosity levels, I didn't think it was worth fixing. However, we could conceivably wrap every call to `reportSDoc` in a `localTCState` if it becomes a problem. (or something more fine-grained that just restores the shadowing information)

> Dunno, I liked the old error better, because it prints what the user wrote. The user can still not see which of the As has been renamed to A_1.

The parameter `A` is not subject to renaming, so there is only one thing it could be.

andreasabel committed:
[ #4697 #4699 ] work around state messing with printing

Any invocation of prettyTCM can in principle jinx the printer.

Jesper explains that prettyTCM registers the module parameters as being
used, which might result in renaming of bound variables further down the
pipeline.

andreasabel commented:
For now, I abstain from `prettyTCM` here to restore the original printing.
If the original printing is considered wrong, there should be a systematic look at it.  (Certainly my change should not change the printing!)
